wurst/lib/ItemRecipe.wurst

```diff
@@ -0,0 +1,570 @@
+        if batch
+            batch = false
+            count++
+
+    /** Sets or removes specified ability from triggering recipe assembly. */
+    function setAbility(int abilityId)
+        LinkedList<ItemRecipe> recipes
+
+        if this.abilityId == abilityId
+            return
+
+        if this.abilityId != 0
+            recipes = recipeMap.get(abilityId)
+            recipes.remove(this)
+            if recipes.isEmpty()
+                destroy recipes
+                recipeMap.remove(abilityId)
+
+        if abilityId > 0
+            this.abilityId = abilityId
+            if not recipeMap.has(abilityId)
+                recipes = new LinkedList<ItemRecipe>()
+                recipes.push(this)
+                recipeMap.put(abilityId, recipes)
+            else
+                recipes = recipeMap.get(abilityId)
+                if not recipes.contains(this)
+                    recipes.push(this)
+
+    /** Removes all entries that match specified item type from recipe ingredient list. */
+    function removeItem(int itemTypeId)
+        let iter = ingredients.iterator()
+        boolean found = false
+
+        while iter.hasNext()
+            if iter.next().itemTypeId == itemTypeId
+                iter.remove()
+                found = true
+        iter.close()
+
+        if found
+            let recipes = recipeMap.get(itemTypeId)
+            recipes.remove(this)
+            if recipes.isEmpty()
+                destroy recipes
+                recipeMap.remove(itemTypeId)
+
+    /** Adds new entry to recipe ingredient list. */
+    function addItem(int itemTypeId, boolean remove, int charges)
+        LinkedList<ItemRecipe> recipes
+        int chrgs = charges
+
+        if itemTypeId <= 0 or itemTypeId == reward
+            return
+        if count >= bj_MAX_INVENTORY and not batch
+            return
+
+        if not recipeMap.has(itemTypeId)
+            recipes = new LinkedList<ItemRecipe>()
+            recipes.push(this)
+            recipeMap.put(itemTypeId, recipes)
+        else
+            recipes = recipeMap.get(itemTypeId)
+            if not recipes.contains(this)
+                recipes.push(this)
+
+        if chrgs < 0
+            chrgs = 0
+
+        let data = new ItemData(itemTypeId, remove, chrgs, count)
+        ingredients.push(data)
+        if not batch
+            count++
+
+    /** Adds new entry to recipe ingredient list. */
+    function addItem(int itemTypeId)
+        addItem(itemTypeId, true, 0)
+
+    private function orderedSearch(LinkedList<item> items) returns HashMap<int, ItemData>
+        int slot = 0
+        boolean found = false
+        item itm
+        int charges
+        ItemData data
+        int idx
+        let recipeData = new HashMap<int, ItemData>()
+
+        let iter = ingredients.iterator()..next()
+        while iter.current != iter.dummy // Match() validated count > size already
+            itm = items.get(slot)
+            charges = GetItemCharges(itm)
```

Comment:
Maybe we should put up a PR to add a `itm.getCharges()` method to the public library

Reply:
I'm in contact with Frotty quite often lately. I'll be moving few of these to wurstStdlib2 if allowed. I'm already a contributor to wurst 😀